src/qtism/runtime/common/VariableFactoryInterface.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
+ *
+ * @author Sergei Mikhailov <sergei.mikhailov@taotesting.com>
+ * @license GPLv2
+ */
+
+declare(strict_types=1);
+
+namespace qtism\runtime\common;
+
+use UnexpectedValueException;
+use qtism\data\state\VariableDeclaration;
+
+interface VariableFactoryInterface
```

Comment:
I understand the principle of having an interface for each class (you made the implementation final), but I really don't see another implementation to be necessary anytime soon. I'd drop the interface.

Reply:
I prefer having `final` default implementations exactly to enforce library users not to abuse class extension, which wouldn't be possible without an interface.